### PR TITLE
Integrate nvalchemi-toolkit-ops and new model format

### PR DESCRIPTION
**Pull request by zubatyuk**
## Summary

This PR represents a significant architectural update to the AIMNet2 calculator, integrating the `nvalchemi-toolkit-ops` library for optimized neighbor list construction and dispersion calculations. It also introduces a modular model format that externalizes long-range interactions, allowing for runtime configuration of electrostatic methods without requiring model re-compilation.

**Key changes:**

- **Accelerated Warp kernels**: 
  - Implements custom CUDA kernels for `conv_sv_2d_sp` operations using NVIDIA Warp.
  - Significantly improves performance for sparse convolutions on GPU.

- **nvalchemi-toolkit-ops integration**: 
  - Implements `AdaptiveNeighborList` with automatic buffer resizing and hysteresis to maintain optimal GPU utilization (~75%) while preventing overflow.
  - Integrates optimized DFT-D3 dispersion kernels.

- **Batched periodic systems**: 
  - Adds full support for batched periodic boundary conditions (PBC).
  - Enables stress tensor calculation for periodic systems, facilitating cell optimization and NPT simulations.

- **External long-range modules**: 
  - Decouples `LRCoulomb` and `DFTD3` modules from the core neural network model.
  - Allows runtime switching between electrostatic methods (Simple, Damped Shifted Force, Ewald Summation) via the calculator API.
  - Supports separate neighbor lists for short-range and long-range interactions when cutoffs differ significantly (>20%).

- **New model format**: 
  - Introduces a new v2 format for exported models (PyTorch native `.pt`), containing the model state dict, embedded YAML configuration, and metadata.
  - Provides seamless backward compatibility for legacy JIT-compiled `.jpt` models.
  - Includes CLI tools for model conversion and export.

### CODE REVIEW

**Comment by isayev** [comment]
## PyTorch/GPU Programming Analysis Report

### Executive Summary

This PR introduces significant GPU/PyTorch modernization: custom NVIDIA Warp kernels, `AdaptiveNeighborList` with automatic buffer management, and GPU-accelerated DFT-D3 via `nvalchemiops`. Overall, the code follows modern PyTorch best practices with some issues to address.

---

### Critical Findings

| Priority | Issue | Location |
|----------|-------|----------|
| **High** | AMP incompatibility - Warp kernels hardcoded to float32, may fail under `torch.autocast` | `conv_sv_2d_sp_wp.py` |
| **High** | GPU-CPU sync in hot path - `num_neighbors.max().item()` causes sync every MD step | `AdaptiveNeighborList` |
| **High** | Redundant `idx.to(torch.int32)` creates temporaries on every kernel launch | Warp kernel calls |
| **Medium** | Unfused double backward - 4 sequential kernel launches could be reduced | backward kernels |
| **Medium** | No memory pool - dynamic neighbor list resizing may cause fragmentation | long MD runs |
| **Low** | Dead code: `if _m == M: break` unreachable in kernel loops | Warp kernels |

---

### Highlights (Good Practices)

1. **Modern PyTorch Integration**
   - Uses `@torch.library.custom_op` for PyTorch 2.x compatibility
   - Registers fake tensors via `@torch.library.register_fake` enabling `torch.compile`
   - Proper autograd registration with `@torch.compiler.allow_in_graph`

2. **AdaptiveNeighborList Hysteresis**
   - Shrink at 50% utilization, grow at overflow
   - Target 75% utilization balances memory vs recomputation
   - Aligned allocations (multiple of 16) improve cache efficiency

3. **CPU Fallback**
   - Graceful fallback to pure PyTorch einsum when Warp unavailable
   - No functionality loss, only performance difference

4. **Batched PBC**
   - Correct stress tensor formula (`dE/dcell / volume`)
   - Efficient einsum usage for batched matrix multiplications

---

### Compatibility

| Requirement | Version |
|-------------|---------|
| PyTorch | ≥2.4 |
| CUDA | 11.0+ |
| Compute Capability | 5.0+ (Maxwell) |
| Warp | 1.11+ |

---

### Detailed Analysis

#### 1. Warp Kernel Memory Access

The forward kernel uses indirect memory access via neighbor indices:
```python
a_val = a[_idx, _a, _g]  # INDIRECT ACCESS - potential cache miss
```
This is inherent to sparse neighbor-based operations but impacts bandwidth efficiency. The backward kernel uses atomic operations (`wp.atomic_add`) for gradient accumulation which can cause contention.

#### 2. Tensor Layout

Good practice: explicitly ensures contiguous memory before passing to Warp kernels:
```python
grad_output_contig = grad_output.contiguous()
```

#### 3. AdaptiveNeighborList Buffer Strategy

- Initial buffer: `density * (4/3 * π * cutoff³)`
- Hysteresis band: 50%-100% (no resizing occurs, prevents thrashing)
- Minimum buffer: 16 neighbors

#### 4. Recommendations

1. Add dtype checking for AMP compatibility or implement fp16 kernel variants
2. Batch utilization updates across multiple steps to reduce GPU-CPU sync
3. Store `idx` as `int32` upstream to avoid repeated conversions
4. Consider memory pool for neighbor list buffers in long MD runs
